### PR TITLE
[seraphis_wallet]: encrypted file

### PR DESCRIPTION
**Pull request by ghostway0**
@rbrunner7

### CODE REVIEW

File: src/seraphis_wallet/encrypted_file.h [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    if (!::serialization::serialize(file_ar, file))
+        return false;
+
+    return epee::file_io_utils::save_string_to_file(path, file_oss.str());
+}
+
+template <class T> bool write_encrypted_file_json(std::string path, const crypto::chacha_key &key, T &struct_in)
+{
+    std::string struct_json = epee::serialization::store_t_to_json(struct_in);
+    std::string data = epee::string_encoding::base64_encode(struct_json);
```
Comment (vtnerd) [inline src/seraphis_wallet/encrypted_file.h:102]: Why the base64? The JSON can be encrypted directly?

File: src/seraphis_wallet/encrypted_file.h [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+    encrypted_file file = {};
+    file.iv = crypto::rand<crypto::chacha_iv>();
+
+    std::string encrypted_data;
+    encrypted_data.resize(data.size());
+
+    crypto::chacha20(data.data(), data.size(), key, file.iv, &encrypted_data[0]);
+
+    file.encrypted_data = encrypted_data;
```
Comment (vtnerd) [inline src/seraphis_wallet/encrypted_file.h:112]: std::move()

File: src/seraphis_wallet/encrypted_file.h [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+    encrypted_file file = {};
+    file.iv = crypto::rand<crypto::chacha_iv>();
+
+    std::string encrypted_data;
+    encrypted_data.resize(buf.size());
+
+    crypto::chacha20(buf.data(), buf.size(), key, file.iv, &encrypted_data[0]);
+
+    file.encrypted_data = encrypted_data;
```
Comment (vtnerd) [inline src/seraphis_wallet/encrypted_file.h:89]: std::move()

File: src/seraphis_wallet/encrypted_file.h [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+    binary_archive<false> file_ar{epee::strspan<std::uint8_t>(buf)};
+    if (!::serialization::serialize(file_ar, file))
+        return false;
+
+    std::string decrypted_data;
+    decrypted_data.resize(file.encrypted_data.size());
+    crypto::chacha20(file.encrypted_data.data(), file.encrypted_data.size(), key, file.iv, &decrypted_data[0]);
+
+    binary_archive<false> ar{epee::strspan<std::uint8_t>(decrypted_data)};
```
Comment (vtnerd) [inline src/seraphis_wallet/encrypted_file.h:44]: I would consider using the epee binary output. It is more flexible when adding new fields than this cryptonote scheme (which is more compact).

Comment (rbrunner7) [reply src/seraphis_wallet/encrypted_file.h:44]: > using the epee binary output

Which class is this exactly?

The whole `wallet2` code is pretty much standardized on `binary_archive` now, especially since moneromooo deprecated the use of the outdated Boost serialization a few years back. I think `binary_archive` has proven itself quite well over all that time, and would hesitate to give up the standardization.

File: src/seraphis_wallet/encrypted_file.h [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+    binary_archive<false> file_ar{epee::strspan<std::uint8_t>(buf)};
+    if (!::serialization::serialize(file_ar, file))
+        return false;
+
+    std::string decrypted_data;
+    decrypted_data.resize(file.encrypted_data.size());
+    crypto::chacha20(file.encrypted_data.data(), file.encrypted_data.size(), key, file.iv, &decrypted_data[0]);
+
+    std::string decoded = epee::string_encoding::base64_decode(decrypted_data);
```
Comment (rbrunner7) [inline src/seraphis_wallet/encrypted_file.h:68]: Like @vtnerd I don't see any good argument for Base64 here, but maybe you had one in mind, or you copied this from somewhere where we could then check whether the conditions / the use case are really similar or not.

File: tests/unit_tests/encrypted_file.cpp [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    END_SERIALIZE()
+
+    BEGIN_KV_SERIALIZE_MAP()
+    KV_SERIALIZE_VAL_POD_AS_BLOB_FORCE(data)
+    END_KV_SERIALIZE_MAP()
+};
+
+TEST(EncryptedFile, ReadWriteBlob)
+{
+    boost::filesystem::path temp_file = boost::filesystem::unique_path();
```
Comment (rbrunner7) [inline tests/unit_tests/encrypted_file.cpp:24]: Other test cases seem to use the following construct because the method you use here only returns a file name, and then maybe it's not clear *where* the file gets created, but I must confess that I don't know much about that stuff:

     boost::filesystem::temp_directory_path() / boost::filesystem::unique_path()

File: tests/unit_tests/key_container.cpp
Comment (rbrunner7) [inline tests/unit_tests/key_container.cpp:1]: This file should probably not be part of this PR?

**Review by rbrunner7 — COMMENTED** [review]
A general question for the PR as a whole: As this will probably be merged first, and be the first file in the not-yet-existing */src/seraphis_wallet* folder, doesn't this need some *CMakeList.txt* file as well to be self-contained, create the folder, and start some library?

Maybe it will take some time to get the key container merge-ready, and therefore a good idea to have a state where other people could already add more code.

File: src/seraphis_wallet/encrypted_file.h (hunk 4 above)
Comment (vtnerd) [reply src/seraphis_wallet/encrypted_file.h:44]: > Which class is this exactly?


> The whole wallet2 code is pretty much standardized on binary_archive now

I forgot about that. The format is more compact, but you have to remember to add a version field manually in most cases (or some type field), as the field ordering is strict and implied.

The epee binary format is close to msgpack in terms of performance and flexibility (especially when adding new fields).

File: src/seraphis_wallet/encrypted_file.h (hunk 5 above)
Comment (ghostway0) [reply src/seraphis_wallet/encrypted_file.h:68]: I don't remember why I did that, I think the argument was in a meeting somewhere that it was easier to copy?
will change it back

File: tests/unit_tests/key_container.cpp
Comment (ghostway0) [reply tests/unit_tests/key_container.cpp:1]: locally I removed it, probably didn't push?

File: src/seraphis_wallet/encrypted_file.h (hunk 4 above)
Comment (rbrunner7) [reply src/seraphis_wallet/encrypted_file.h:44]: Let us stay with the same stuff that `wallet2` also uses. Using the same serialization system for the file as a whole as for individual classes seems ok to me for the time being.

**Comment by rbrunner7** [comment]
> I preferred to let this be in the key container PR because it would add a (mostly) empty file, although, the key container seems to still be a bit raw from the comments so maybe

So be it. I still would have preferred this to be nicely self-contained by adding a *CMakeList.txt* file, but the code in this PR waited already long enough. We have to get things moving.

**Review by rbrunner7 — APPROVED** [review]
Looks ok to me after a round of improvements.
